temporal-sdk/src/main/java/io/temporal/internal/worker/StickyQueueBalancer.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import io.temporal.api.enums.v1.TaskQueueKind;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.concurrent.ThreadSafe;
+
+@ThreadSafe
+public class StickyQueueBalancer {
+  private final int pollersCount;
+  private final boolean stickyQueueEnabled;
+  private final AtomicInteger stickyPollers = new AtomicInteger(0);
+  private final AtomicInteger normalPollers = new AtomicInteger(0);
+
+  private volatile long stickyBacklogSize = 0;
+
+  public StickyQueueBalancer(int pollersCount, boolean stickyQueueEnabled) {
+    this.pollersCount = pollersCount;
+    this.stickyQueueEnabled = stickyQueueEnabled;
+  }
+
+  /**
+   * @return task queue kind that should be used for the next poll
+   */
+  public TaskQueueKind makePoll() {
+    if (stickyQueueEnabled) {
+      // If pollersCount >= stickyBacklogSize > 0 we want to go back to a normal ratio to avoid a
+      // situation that too many pollers (all of them in the worst case) will open only sticky queue
+      // polls observing a stickyBacklogSize == 1 for example (which actually can be 0 already at
+      // that
+      // moment) and get stuck causing dip in worker load.
+      if (stickyBacklogSize > pollersCount || stickyPollers.get() <= normalPollers.get()) {
```

Comment:
I assume it is ok that this conditional is not atomic as a whole? Meaning, it's ok if _technically_ in a racy situation this results in an unexpected sticky or normal poll when `finishPoll` is run by another thread at the same time? Otherwise, you'd be better off switching the atomic integers to regular ones and marking these methods `synchronized` (calls shouldn't be so frequent and time sensitive here that `synchronized` would cause a real problem IMO).

Reply:
Yeah, I think it's ok here. And the main argument is that the value of stickyBacklogSize is anyway non-precise. We don't control/verify which order the responses will be coming in and that the value that we write into stickyBacklogSize is an actual last value. In this condition, it's not worth fighting for precise decisions here.